core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2517,6 +2517,18 @@ public static Object item(Object object, Object index) {
     if (object instanceof List && index instanceof Number) {
       return arrayItem((List) object, ((Number) index).intValue());
     }
+    if (index instanceof Number) {
```

Comment:
According to the signature of the operator it should be an Integer, not any number, is that right?

Reply:
Yes, you are right, in the current implementation I have favoured consistency with ARRAY (just above) which has `<INTEGER>` as index signature, but nonetheless accepts any number and converts it to an integer.

In case we want to follow strictly the signature and just support INTEGER, I propose we do the same for ARRAY (i.e., we test `index instanceof Integer`), so we don't have surprising effects depending on the field data type.

What do you think?

Reply:
I guess that if everything passes from the validator then we should never arrive here with something that is not an integer but to be safe we could add an assertion or `IllegalStateException` to be a bit more defensive.

Reply:
I had a closer look, for both `ARRAY` and `ROW` we require the index to be in the int family (`SqlTypeFamily.INTEGER`), which includes `TINYINT`, `SMALLINT`, `INTEGER` and `BIGINT`, so we can't really throw if the index is not `int`.

I have tried other numeric values not in the int family, and the validator fails, when we reach this point type validation already guarantees we have a correct type, and relying on `Numeric` becomes just a shorthand for `Long`, `Short`, `Integer`, etc.

If you agree I'd be inclined to leave it as-is, since adding an assert for each of the legal types would harm readability for not much.

Reply:
Thanks for the clarification. Yes let's leave it as is.